### PR TITLE
Added the implementation to automatically add the winrm security rule for an azure security group

### PR DESCRIPTION
**Pull request by pavanadepu2**

### CODE REVIEW

**Comment by niadak** [comment]
Looks fine

File: Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1 [hunk 1]
```diff
@@ -340,4 +340,108 @@ function Remove-AzureMachineCustomScriptExtension
...
+        $securityGroupName = $securityGroup.Name
+        try
+        {
+            Write-Verbose -Verbose "[Azure Call]Getting network security rule config $ruleName under security group $securityGroupName"
+            $winRMConfigRule = Get-AzureRmNetworkSecurityRuleConfig -NetworkSecurityGroup $securityGroup -Name $ruleName -EA SilentlyContinue
+            Write-Verbose -Verbose "[Azure Call]Got network security rule config $ruleName under security group $securityGroupName"
+        }
+        catch
+        {
+            #Ignore the exception
```
Comment (kmkumaran) [inline Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1]: Should you return right here? Else you will proceed further to print message that rule already exists.

File: Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1 [hunk 2]
```diff
@@ -340,4 +340,108 @@ function Remove-AzureMachineCustomScriptExtension
...
+        {            
+            for($i=1; $i -le 3; $i++)
+            {
+                try
+                {
+                    Write-Verbose -Verbose "[Azure Call]Adding inbound network security rule config $ruleName with priority $rulePriotity for port $winrmHttpsPort under security group $securityGroupName"
+                    $result = Add-AzureRmNetworkSecurityRuleConfig -NetworkSecurityGroup $securityGroup -Name $ruleName -Direction Inbound -Access Allow -SourceAddressPrefix '*' -SourcePortRange '*' -DestinationAddressPrefix '*' -DestinationPortRange $winrmHttpsPort -Protocol * -Priority $rulePriotity | Set-AzureRmNetworkSecurityGroup
+                    Write-Verbose -Verbose "[Azure Call]Added inbound network security rule config $ruleName with priority $rulePriotity for port $winrmHttpsPort under security group $securityGroupName"                         
+                }
+                catch
```
Comment (kmkumaran) [inline Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1]: Is there a specific exception that says rule priority is conflicting?

File: Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1 [hunk 3]
```diff
@@ -340,4 +340,108 @@ function Remove-AzureMachineCustomScriptExtension
...
+                {           
+                    $newPort = [convert]::ToInt32($rulePriotity, 10) + 50;
```
Comment (kmkumaran) [inline Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1]: Should this responsibility be with caller? In other words should this method take a list of ports and attempt to apply them. You can then have flexibility in callers to pass any list of port that they think is right for adding. In this case, Azure file copy can pass 3 choices with + 50 difference. Thoughts?

File: Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1 [hunk 4]
```diff
@@ -340,4 +340,108 @@ function Remove-AzureMachineCustomScriptExtension
...
+{
+     param([string] $resourceGroupName,
+           [string] $vmId)
+
+    $securityGroup = $null
+
+    if(-not [string]::IsNullOrEmpty($resourceGroupName) -and -not [string]::IsNullOrEmpty($vmId))
+    {
+        Write-Verbose -Verbose "[Azure Call]Getting network interfaces in resource group $resourceGroupName for vm $vmId"
+        $networkInterface = Get-AzureRmNetworkInterface -ResourceGroupName $resourceGroupName | Where-Object { $_.VirtualMachine.Id -eq $vmId }
```
Comment (kmkumaran) [inline Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1]: Can this be a list? I know you are not dealing with multiple nics. But if there is any VM returning multiple nic, you might want to gracefully exit.

File: Tasks/DeployAzureResourceGroup/Utility.ps1 [hunk 5]
```diff
@@ -896,25 +923,28 @@ function Add-AzureVMCustomScriptExtension
...
-    if($isExtensionExists)
-    {
-        Write-Verbose -Verbose "Skipping the addition of custom script extension '$extensionName' as it already exists"
-        return
-    }
+        if($isExtensionExists)
+        {
+            Write-Verbose -Verbose "Skipping the addition of custom script extension '$extensionName' as it already exists"
+
+            Add-WinRMHttpsNetworkSecurityRuleConfig -resourceGroupName $resourceGroupName -vmId $vmId -ruleName $ruleName -rulePriotity $rulePriotity -winrmHttpsPort $winrmHttpsPort
```
Comment (kmkumaran) [inline Tasks/DeployAzureResourceGroup/Utility.ps1]: I am missing this part. If extension exists, why do you call securityruleconfig? Can't the existing extension be for machinecustomscript?

File: Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1 (hunk 1 above)
Comment (pavanadepu2) [inline Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1]: We need the below log message to confirm that the rule already exists. This is useful for debugging in failure cases.

File: Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1 (hunk 3 above)
Comment (pavanadepu2) [inline Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1]: This makes sense if the user is providing the preferred port. But in the current case we are not taking this input from the user, so there would be no difference.

File: Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1 (hunk 2 above)
Comment (pavanadepu2) [inline Tasks/AzureFileCopy/AzureUtilityGTE1.0.ps1]: It is a generic exception with proper message.

File: Tasks/DeployAzureResourceGroup/Utility.ps1 (hunk 5 above)
Comment (pavanadepu2) [inline Tasks/DeployAzureResourceGroup/Utility.ps1]: If the addition of security rule is failed in the first attempt then, it never adds the security group if we miss here.
This is a quick call which is idempotent.

**Comment by kmkumaran** [comment]
No longer relevant. Closing it.

**Pull request closed without merging**